# Association Rules Algorithm

TASK:You are provided with a dataset which contains some data resulting from Mammography tests.
Description of dataset. Mammography is the most effective method for breast cancer screening available today. However, the low positive predictive value of breast biopsy resulting from mammogram interpretation leads to ap- proximately 70% unnecessary biopsies with benign outcomes. To reduce the high number of unnecessary breast biopsies, several computer-aided diagnosis (CAD) systems have been proposed in the last years.These systems help physi- cians in their decision to perform a breast biopsy on a suspicious lesion seen in a mammogram or to perform a short term follow-up examination instead. The dataset we provided can be used to predict the severity (benign or malignant) of a mammographic mass lesion from the attributes of the mass lesion (size, shape, etc.) and the patient’s age. Additionally, the dataset contains for each patient the BI-RADS assessment, which is a score ranging from 0 (definitely benign) to 6 (highly suggestive of malignancy) given by the radiologist upon checking the results of the test. The ground truth is also provided (the severity field), which specifies whether the corresponding mass lesion is benign (0) or malign (1). In the dataset, there are 516 benign and 445 malignant masses that have been identified on full field digital mammograms collected at the Institute of Radiology of the University Erlangen-Nuremberg (Germany) between 2003 and 2006.

## This solution employ mlxtend library.
Mlxtend (machine learning extensions) is a Python library of useful tools for the day-to-day data science tasks. (from Github)

In [14]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

import math
import pandas as pd

In [33]:
# the dataset needs preprocessing to one-hot encoding  
# and remove any NaN values to use apriori class
dataSet = pd.read_csv(r"mammographic_masses.csv",delimiter=",",header=0)
dataList = dataSet.values.tolist()

for i in range(len(dataList)):
    j=0
    while (True):
        if (type (dataList[i][j]) == float and math.isnan(dataList[i][j])):
            del dataList[i][j]
            j -= 1
        j += 1
        if (j>len(dataList[i])-1):
            break

#dataList

### Q1: Report 3 rules with support at least 0.2 and confidence at least 0.9. Specify for each of them the support and the confidence.

In [34]:
# prepare clear demonstration for later output
for i in range (len(dataList)):
    for j in range (len(dataList[i])):
        dataList[i][j] = dataSet.columns[j] + '= ' + str(dataList[i][j])
dataList

[['BI-RADS= 5',
  'Age= 67',
  'Shape= 3',
  'Margin= 5',
  'Density= 3',
  'Severity= 1'],
 ['BI-RADS= 4',
  'Age= 43',
  'Shape= 1',
  'Margin= 1',
  'Density= ?',
  'Severity= 1'],
 ['BI-RADS= 5',
  'Age= 58',
  'Shape= 4',
  'Margin= 5',
  'Density= 3',
  'Severity= 1'],
 ['BI-RADS= 4',
  'Age= 28',
  'Shape= 1',
  'Margin= 1',
  'Density= 3',
  'Severity= 0'],
 ['BI-RADS= 5',
  'Age= 74',
  'Shape= 1',
  'Margin= 5',
  'Density= ?',
  'Severity= 1'],
 ['BI-RADS= 4',
  'Age= 65',
  'Shape= 1',
  'Margin= ?',
  'Density= 3',
  'Severity= 0'],
 ['BI-RADS= 4',
  'Age= 70',
  'Shape= ?',
  'Margin= ?',
  'Density= 3',
  'Severity= 0'],
 ['BI-RADS= 5',
  'Age= 42',
  'Shape= 1',
  'Margin= ?',
  'Density= 3',
  'Severity= 0'],
 ['BI-RADS= 5',
  'Age= 57',
  'Shape= 1',
  'Margin= 5',
  'Density= 3',
  'Severity= 1'],
 ['BI-RADS= 5',
  'Age= 60',
  'Shape= ?',
  'Margin= 5',
  'Density= 1',
  'Severity= 1'],
 ['BI-RADS= 5',
  'Age= 76',
  'Shape= 1',
  'Margin= 4',
  'Density= 3',
  'Sev

In [35]:
# create a Boolean list for apriori algorithm
te = TransactionEncoder()
teArray = te.fit(dataList).transform(dataList)
dataF = pd.DataFrame(teArray, columns = te.columns_)

In [36]:
# calculate frequent itemsets of 0.2 support and 0.9 confidence
frequentItemSets = apriori (dataF, min_support=0.2, use_colnames=True)
resultAR = association_rules (frequentItemSets, metric='confidence', min_threshold=0.9)
resultAR [["antecedents","consequents","support","confidence"]]

antecedents    consequents   support  confidence
0                             (Shape= 4)   (Density= 3)  0.374610    0.900000
1                (Margin= 1, Density= 3)   (BI-RADS= 4)  0.263267    0.900356
2               (Margin= 1, Severity= 0)   (BI-RADS= 4)  0.299688    0.911392
3                (Margin= 1, BI-RADS= 4)  (Severity= 0)  0.299688    0.911392
4                 (Shape= 4, BI-RADS= 5)   (Density= 3)  0.245578    0.904215
5                 (Shape= 4, BI-RADS= 5)  (Severity= 1)  0.246618    0.908046
6                (Shape= 4, Severity= 1)   (Density= 3)  0.295525    0.901587
7   (Margin= 1, Density= 3, Severity= 0)   (BI-RADS= 4)  0.238293    0.927126
8    (Margin= 1, Density= 3, BI-RADS= 4)  (Severity= 0)  0.238293    0.905138
9     (Shape= 4, BI-RADS= 5, Density= 3)  (Severity= 1)  0.224766    0.915254
10   (Shape= 4, BI-RADS= 5, Severity= 1)   (Density= 3)  0.224766    0.911392

### Anwser to Q1:

From the sheet above, we can see the first three rules are:

Shape=4 -> Density=3 has 0.37 support and 0.9 confidence.

Margin=1, Density=3 -> BI-RADS=4 has 0.26 support and 0.9 confidence.

Marigin=1, Severity=0 -> BI-RADS=4 has 0.3 support and  0.91 confidence.

### Q2: Report one or two rules with the specified requirements that you think might help us predicting whether a given instance is benign or malign. You should not report rules with the attribute BI-RADS for this question. Which insights did you get from those rules? (e.g. the margin of the lesion can help us determining whether a lesion is benign or malign).

Remember that only rules with support at least 0.1, (i.e. their frequency is at least 10%) are relevant for us. Rules with lower support are usually not informative, as there is no much evidence they are true or not. In our exercise we consider relevant any rule with confidence at least 0.9 (i.e. they are true 90% of times).

In [40]:
frequentItemSets2 = apriori (dataF, min_support=0.1, use_colnames=True)
resultAR2 = association_rules (frequentItemSets2, metric='confidence', min_threshold=0.9)
resultAR2 [["antecedents","consequents","support","confidence"]]

antecedents    consequents   support  \
0                                       (Margin= 5)   (Density= 3)  0.131113   
1                                        (Shape= 4)   (Density= 3)  0.374610   
2                           (Margin= 1, Density= 3)   (BI-RADS= 4)  0.263267   
3                          (Margin= 1, Severity= 0)   (BI-RADS= 4)  0.299688   
4                           (Margin= 1, BI-RADS= 4)  (Severity= 0)  0.299688   
5                            (BI-RADS= 4, Shape= 1)  (Severity= 0)  0.173777   
6                            (BI-RADS= 4, Shape= 2)  (Severity= 0)  0.162331   
7                            (Shape= 4, BI-RADS= 5)   (Density= 3)  0.245578   
8                            (Shape= 4, BI-RADS= 5)  (Severity= 1)  0.246618   
9                          (Severity= 1, Margin= 5)   (Density= 3)  0.110302   
10                            (Shape= 4, Margin= 5)   (Density= 3)  0.112383   
11                          (Shape= 4, Severity= 1)   (Density= 3)  0.295525   
12                            (Margin= 1, Shape= 2)  (Severity= 0)  0.136316   
13             (Margin= 1, Density= 3, Severity= 0)   (BI-RADS= 4)  0.238293   
14              (Margin= 1, Density= 3, BI-RADS= 4)  (Severity= 0)  0.238293   
15                (Margin= 1, Density= 3, Shape= 1)   (BI-RADS= 4)  0.144641   
16               (Density= 3, BI-RADS= 4, Shape= 1)    (Margin= 1)  0.144641   
17              (Density= 3, Severity= 0, Shape= 1)   (BI-RADS= 4)  0.145682   
18               (Density= 3, BI-RADS= 4, Shape= 1)  (Severity= 0)  0.145682   
19              (Density= 3, Severity= 0, Shape= 2)   (BI-RADS= 4)  0.121748   
20               (Margin= 1, Severity= 0, Shape= 1)   (BI-RADS= 4)  0.156087   
21                (Margin= 1, BI-RADS= 4, Shape= 1)  (Severity= 0)  0.156087   
22               (Margin= 1, Severity= 0, Shape= 2)   (BI-RADS= 4)  0.124870   
23                (Margin= 1, BI-RADS= 4, Shape= 2)  (Severity= 0)  0.124870   
24              (Margin= 4, BI-RADS= 5, Density= 3)  (Severity= 1)  0.121748   
25               (Shape= 4, BI-RADS= 5, Density= 3)  (Severity= 1)  0.224766   
26              (Shape= 4, BI-RADS= 5, Severity= 1)   (Density= 3)  0.224766   
27                (Shape= 4, Margin= 4, BI-RADS= 5)  (Severity= 1)  0.109261   
28                (Margin= 1, Density= 3, Shape= 1)  (Severity= 0)  0.143600   
29              (Density= 3, Severity= 0, Shape= 1)    (Margin= 1)  0.143600   
30   (Margin= 1, Density= 3, Severity= 0, Shape= 1)   (BI-RADS= 4)  0.133195   
31  (Density= 3, Severity= 0, BI-RADS= 4, Shape= 1)    (Margin= 1)  0.133195   
32    (Margin= 1, Density= 3, BI-RADS= 4, Shape= 1)  (Severity= 0)  0.133195   

    confidence  
0     0.926471  
1     0.900000  
2     0.900356  
3     0.911392  
4     0.911392  
5     0.907609  
6     0.901734  
7     0.904215  
8     0.908046  
9     0.929825  
10    0.955752  
11    0.901587  
12    0.903448  
13    0.927126  
14    0.905138  
15    0.908497  
16    0.908497  
17    0.921053  
18    0.915033  
19    0.914062  
20    0.920245  
21    0.909091  
22    0.916031  
23    0.937500  
24    0.906977  
25    0.915254  
26    0.911392  
27    0.913043  
28    0.901961  
29    0.907895  
30    0.927536  
31    0.914286  
32    0.920863

In [43]:
# eliminates the rows according to the requirements
# consequents should be Severity
# no BI-RADS involved

rightAn = []

for i, row in resultAR2.iterrows():
    if ('Severity= 1' in row['consequents']) or ('Severity= 0' in row['consequents']):
        if ('BI-RADS= 0' not in row['antecedents']) and ('BI-RADS= 1' not in row['antecedents']) and ('BI-RADS= 2' not in row['antecedents']) and ('BI-RADS= 3' not in row['antecedents']) and ('BI-RADS= 4' not in row['antecedents']) and ('BI-RADS= 5' not in row['antecedents']) and ('BI-RADS= 6' not in row['antecedents']):
            rightAn.append(i)
print(rightAn)

[12, 28]


### Anwser to Q2:

From the list, we find 12th and 28th rows meet the requirements.

12	(Margin= 1, Shape= 2) -> (Severity= 0)	support=0.136316	confidence=0.903448

28	(Margin= 1, Density= 3, Shape= 1) -> (Severity= 0)	support=0.143600	confidence=0.901961

#### Insights from the rules:

If a lesion has an oval shape and a circumscribed margin, it is more likely to be benign. Similarly, if a lesion has a round shape, a circumscribed margin, and low density, it is also more likely to be benign.

### Q3: As discussed above, the BI-RADS assessment is not always accurate and it might lead to unnecessary breast biopsy. Provide one or two rules that might give some evidence that the BI-RADS assessment is not always accurate. Explain your answer.

Firsly, assume BI-RADS above 3 is suppose to reveal malignancy. Let's check how offen high BI-RADS relating to benign. To understand why unnecessary breast biopsy exist, we need to check how offen low BI-RAD relating to malignancy as well.

In [51]:
# We use resultAR2 again here

rightAn2 = []
for i, row in resultAR2.iterrows():
    if ('Severity= 0' in row['consequents']):
        if ('BI-RADS= 4' in row['antecedents']) or ('BI-RADS= 5' in row['antecedents']) or ('BI-RADS= 6' in row['antecedents']):
            rightAn2.append(i)

resultAR2.loc[rightAn2, ["antecedents","consequents","support","confidence"]]

antecedents    consequents   support  \
4                         (Margin= 1, BI-RADS= 4)  (Severity= 0)  0.299688   
5                          (BI-RADS= 4, Shape= 1)  (Severity= 0)  0.173777   
6                          (BI-RADS= 4, Shape= 2)  (Severity= 0)  0.162331   
14            (Margin= 1, Density= 3, BI-RADS= 4)  (Severity= 0)  0.238293   
18             (Density= 3, BI-RADS= 4, Shape= 1)  (Severity= 0)  0.145682   
21              (Margin= 1, BI-RADS= 4, Shape= 1)  (Severity= 0)  0.156087   
23              (Margin= 1, BI-RADS= 4, Shape= 2)  (Severity= 0)  0.124870   
32  (Margin= 1, Density= 3, BI-RADS= 4, Shape= 1)  (Severity= 0)  0.133195   

    confidence  
4     0.911392  
5     0.907609  
6     0.901734  
14    0.905138  
18    0.915033  
21    0.909091  
23    0.937500  
32    0.920863

In [52]:
rightAn3 = []
for i, row in resultAR2.iterrows():
    if ('Severity= 1' in row['consequents']):
        if ('BI-RADS= 0' in row['antecedents']) or ('BI-RADS= 1' in row['antecedents']) or ('BI-RADS= 2' in row['antecedents']):
            rightAn3.append(i)

resultAR2.loc[rightAn3, ["antecedents","consequents","support","confidence"]]

Empty DataFrame
Columns: [antecedents, consequents, support, confidence]
Index: []

### Anwser to Q3:

From the result above, we can find BI-RADS=4 appearing a lot in antecedents when the consequents is Severity=0. The listed association rules all have support over 0.1 and confidence over 0.9. Based on the results, we can assume that when BI-RADS = 4, it often leads to UNNECESSARY biopsies with benign outcomes. 


However, we can't dismiss the instructive nature of BI-RADS because through data mining, we can see that low BI-RADS value and Severity=1 are never strongly correlated. So this indicator hardly leads to missed diagnosis of malignant chest tumors, for example.
We need to further improve the association rules by data mining on the current basis to reduce unnecessary biopsies.

### Q4: Write a script in Python to find the confidence and support of the following rule: Age=35 ⇒ Severity=0. Report its support and confidence. Do you think this rule tells us something valuable or that we should ignore it as there is not enough evidence to support this rule?

In [53]:
# the support and confidence are unknow, therefore we lower the threshold
frequentItemSets4 = apriori (dataF, min_support=0.0001, use_colnames=True)
resultAR4 = association_rules (frequentItemSets4, metric='confidence', min_threshold=0)
resultAR4 [["antecedents","consequents","support","confidence"]]

antecedents                                        consequents  \
0           (Age= 18)                                       (BI-RADS= 4)   
1        (BI-RADS= 4)                                          (Age= 18)   
2        (Density= 3)                                          (Age= 18)   
3           (Age= 18)                                       (Density= 3)   
4         (Margin= 1)                                          (Age= 18)   
...               ...                                                ...   
241307     (Shape= 4)  (Margin= 4, Age= ?, BI-RADS= 5, Severity= 1, D...   
241308       (Age= ?)  (Margin= 4, Shape= 4, BI-RADS= 5, Severity= 1,...   
241309   (BI-RADS= 5)  (Margin= 4, Shape= 4, Age= ?, Severity= 1, Den...   
241310  (Severity= 1)  (Margin= 4, Shape= 4, Age= ?, BI-RADS= 5, Dens...   
241311   (Density= 3)  (Margin= 4, Shape= 4, Age= ?, BI-RADS= 5, Seve...   

         support  confidence  
0       0.001041    1.000000  
1       0.001041    0.001828  
2       0.001041    0.001253  
3       0.001041    1.000000  
4       0.001041    0.002801  
...          ...         ...  
241307  0.002081    0.005000  
241308  0.002081    0.400000  
241309  0.002081    0.005780  
241310  0.002081    0.004494  
241311  0.002081    0.002506  

[241312 rows x 4 columns]

In [55]:
rightAn4 = []

for i, row in resultAR4.iterrows():
    if ('Severity= 0' in row['consequents']) and (len(row['consequents'])==1):
        if ('Age= 35' in row['antecedents']) and (len(row['antecedents'])==1):
            rightAn4.append(i)
            break

resultAR4.loc[rightAn4, ["antecedents","consequents","support","confidence"]]

antecedents    consequents   support  confidence
301   (Age= 35)  (Severity= 0)  0.012487    0.923077

### Anwser to Q4:

The support of Age=35 -> Severity=0 is 0.012, while the confidence is 0.92. Although the confidence is high in this rule, the support not high enough, which means that not enough data category under Age=35 to draw a solid conclusion. 

### Q5: Provide at least one rule of that kind with support at least 0.1 and confidence at least 0.9.

The attribute “Age” is ordinal which makes the rule mining approach not ideal. In particular, one would like to obtain rules of the kind
Age≥ n,A1 = a1,...,Ak = ak − > Severity=’1’
(where n is an integer), as the age is an important factor in determining whether a given instance is malign or benign. However, this issue can be circumvented in our case by modifying the input file (the ’csv’ file) accordingly. Be careful on how you handle the missing values (i.e. those with a ’?’).

In [77]:
# processing the age attribute to nominal
# '?' is replaced by NaN, as replacing '?' with certain numbers may arise wrong association rules

newData = pd.read_csv(r"mammographic_masses.csv",delimiter=",",header=0, na_values='?')
ages = list(age for age in newData['Age'].unique() if not math.isnan(age))

# we use 40s,50s,60s etc. to category the ordinal ages to nominal
age10 = [age for age in range (100,0,-10)]
#ages,age10

In [87]:
ageGroup = []
for i, row in newData.iterrows():
    ageGroup_row = []
    for var in row.index:
        # handle NaN
        if math.isnan(row[var]):
            continue
        if var == 'Age':
            for i in age10:
                if row[var] >= i:
                    ageGroup_row.append(var + '= ' + str(i) + 's')
                    break
        else:
            ageGroup_row.append(var + '= ' +str(row[var]))
    ageGroup.append(ageGroup_row)
    
ageGroup

[['BI-RADS= 5.0',
  'Age= 60s',
  'Shape= 3.0',
  'Margin= 5.0',
  'Density= 3.0',
  'Severity= 1.0'],
 ['BI-RADS= 4.0', 'Age= 40s', 'Shape= 1.0', 'Margin= 1.0', 'Severity= 1.0'],
 ['BI-RADS= 5.0',
  'Age= 50s',
  'Shape= 4.0',
  'Margin= 5.0',
  'Density= 3.0',
  'Severity= 1.0'],
 ['BI-RADS= 4.0',
  'Age= 20s',
  'Shape= 1.0',
  'Margin= 1.0',
  'Density= 3.0',
  'Severity= 0.0'],
 ['BI-RADS= 5.0', 'Age= 70s', 'Shape= 1.0', 'Margin= 5.0', 'Severity= 1.0'],
 ['BI-RADS= 4.0', 'Age= 60s', 'Shape= 1.0', 'Density= 3.0', 'Severity= 0.0'],
 ['BI-RADS= 4.0', 'Age= 70s', 'Density= 3.0', 'Severity= 0.0'],
 ['BI-RADS= 5.0', 'Age= 40s', 'Shape= 1.0', 'Density= 3.0', 'Severity= 0.0'],
 ['BI-RADS= 5.0',
  'Age= 50s',
  'Shape= 1.0',
  'Margin= 5.0',
  'Density= 3.0',
  'Severity= 1.0'],
 ['BI-RADS= 5.0', 'Age= 60s', 'Margin= 5.0', 'Density= 1.0', 'Severity= 1.0'],
 ['BI-RADS= 5.0',
  'Age= 70s',
  'Shape= 1.0',
  'Margin= 4.0',
  'Density= 3.0',
  'Severity= 1.0'],
 ['BI-RADS= 3.0',
  'Age= 40s',


In [88]:
te2 = TransactionEncoder()
teArray2 = te2.fit(ageGroup).transform(ageGroup)

dataF2 = pd.DataFrame(teArray2, columns = te2.columns_)

In [89]:
frequentIS = apriori (dataF2, min_support=0.1, use_colnames=True)
newAR = association_rules (frequentIS, metric='confidence', min_threshold=0.9)
newAR [["antecedents","consequents","support","confidence"]]

antecedents      consequents  \
0                                       (Margin= 5.0)   (Density= 3.0)   
1                                        (Shape= 4.0)   (Density= 3.0)   
2                            (BI-RADS= 5.0, Age= 60s)   (Density= 3.0)   
3                           (Severity= 1.0, Age= 60s)   (Density= 3.0)   
4                              (Shape= 4.0, Age= 60s)   (Density= 3.0)   
5                         (Density= 3.0, Margin= 1.0)   (BI-RADS= 4.0)   
6                        (Severity= 0.0, Margin= 1.0)   (BI-RADS= 4.0)   
7                         (Margin= 1.0, BI-RADS= 4.0)  (Severity= 0.0)   
8                          (Shape= 1.0, BI-RADS= 4.0)  (Severity= 0.0)   
9                          (Shape= 2.0, BI-RADS= 4.0)  (Severity= 0.0)   
10                         (BI-RADS= 5.0, Shape= 4.0)   (Density= 3.0)   
11                         (BI-RADS= 5.0, Shape= 4.0)  (Severity= 1.0)   
12                       (Severity= 1.0, Margin= 5.0)   (Density= 3.0)   
13                          (Margin= 5.0, Shape= 4.0)   (Density= 3.0)   
14                        (Severity= 1.0, Shape= 4.0)   (Density= 3.0)   
15                          (Shape= 2.0, Margin= 1.0)  (Severity= 0.0)   
16         (Density= 3.0, Severity= 0.0, Margin= 1.0)   (BI-RADS= 4.0)   
17          (Density= 3.0, Margin= 1.0, BI-RADS= 4.0)  (Severity= 0.0)   
18            (Shape= 1.0, Density= 3.0, Margin= 1.0)   (BI-RADS= 4.0)   
19           (Shape= 1.0, Density= 3.0, BI-RADS= 4.0)    (Margin= 1.0)   
20          (Shape= 1.0, Severity= 0.0, Density= 3.0)   (BI-RADS= 4.0)   
21           (Shape= 1.0, Density= 3.0, BI-RADS= 4.0)  (Severity= 0.0)   
22          (Density= 3.0, Severity= 0.0, Shape= 2.0)   (BI-RADS= 4.0)   
23           (Shape= 1.0, Severity= 0.0, Margin= 1.0)   (BI-RADS= 4.0)   
24            (Shape= 1.0, Margin= 1.0, BI-RADS= 4.0)  (Severity= 0.0)   
25           (Severity= 0.0, Shape= 2.0, Margin= 1.0)   (BI-RADS= 4.0)   
26            (Shape= 2.0, Margin= 1.0, BI-RADS= 4.0)  (Severity= 0.0)   
27          (Margin= 4.0, Density= 3.0, BI-RADS= 5.0)  (Severity= 1.0)   
28          (Severity= 1.0, BI-RADS= 5.0, Shape= 4.0)   (Density= 3.0)   
29           (Density= 3.0, BI-RADS= 5.0, Shape= 4.0)  (Severity= 1.0)   
30            (Margin= 4.0, BI-RADS= 5.0, Shape= 4.0)  (Severity= 1.0)   
31          (Shape= 1.0, Severity= 0.0, Density= 3.0)    (Margin= 1.0)   
32            (Shape= 1.0, Density= 3.0, Margin= 1.0)  (Severity= 0.0)   
33  (Shape= 1.0, Severity= 0.0, Density= 3.0, BI-R...    (Margin= 1.0)   
34  (Shape= 1.0, Severity= 0.0, Density= 3.0, Marg...   (BI-RADS= 4.0)   
35  (Shape= 1.0, Density= 3.0, Margin= 1.0, BI-RAD...  (Severity= 0.0)   

     support  confidence  
0   0.131113    0.926471  
1   0.374610    0.900000  
2   0.100937    0.923810  
3   0.131113    0.926471  
4   0.117586    0.949580  
5   0.263267    0.900356  
6   0.299688    0.911392  
7   0.299688    0.911392  
8   0.173777    0.907609  
9   0.162331    0.901734  
10  0.245578    0.904215  
11  0.246618    0.908046  
12  0.110302    0.929825  
13  0.112383    0.955752  
14  0.295525    0.901587  
15  0.136316    0.903448  
16  0.238293    0.927126  
17  0.238293    0.905138  
18  0.144641    0.908497  
19  0.144641    0.908497  
20  0.145682    0.921053  
21  0.145682    0.915033  
22  0.121748    0.914062  
23  0.156087    0.920245  
24  0.156087    0.909091  
25  0.124870    0.916031  
26  0.124870    0.937500  
27  0.121748    0.906977  
28  0.224766    0.911392  
29  0.224766    0.915254  
30  0.109261    0.913043  
31  0.143600    0.907895  
32  0.143600    0.901961  
33  0.133195    0.914286  
34  0.133195    0.927536  
35  0.133195    0.920863

### Anwser to Q4:

3	(Severity= 1.0, Age= 60s)	(Density= 3.0)	0.131113	0.926471

The rule related to age attributes is as follow:
Severity=1.0, Age=60s -> Density=3.0 has support of 0.13 and confidence of 0.93.